near-sdk/src/promise.rs

```diff
@@ -480,6 +480,54 @@ impl Promise {
         other
     }
 
+    /// Schedules execution of multiple concurrent promises right after the
+    /// current promise finishes executing.
+    ///
+    /// This method will send the same return value as a data receipt to all
+    /// following receipts.
+    ///
+    /// In the following code, `bob_near` is created first, `carol_near` second,
+    /// and finally `dave_near` and `eva_near` are created concurrently.
+    ///
+    /// ```no_run
+    /// # use near_sdk::Promise;
+    /// let p1 = Promise::new("bob_near".parse().unwrap()).create_account();
+    /// let p2 = Promise::new("carol_near".parse().unwrap()).create_account();
+    /// let p3 = Promise::new("dave_near".parse().unwrap()).create_account();
+    /// let p4 = Promise::new("eva_near".parse().unwrap()).create_account();
+    /// p1.then(p2).then_concurrent(vec![p3, p4]);
+    /// ```
+    ///
+    /// The returned [`ConcurrentPromises`] allows chaining more promises.
+    ///
+    /// In the following code, `bob_near` is created first, next `carol_near`
+    /// and `dave_near` are created concurrently, and finally `eva_near` is
+    /// created after all others have been created.
+    ///
+    /// ```no_run
+    /// # use near_sdk::Promise;
+    /// let p1 = Promise::new("bob_near".parse().unwrap()).create_account();
+    /// let p2 = Promise::new("carol_near".parse().unwrap()).create_account();
+    /// let p3 = Promise::new("dave_near".parse().unwrap()).create_account();
+    /// let p4 = Promise::new("eva_near".parse().unwrap()).create_account();
+    /// p1.then_concurrent(vec![p2, p3]).join().then(p4);
+    /// ```
+    pub fn then_concurrent(self, promises: Vec<Promise>) -> ConcurrentPromises {
+        let mapped_promises = promises
+            .into_iter()
+            .map(|other| {
+                let self_clone = Promise {
+                    subtype: self.subtype.clone(),
+                    // Cloning `should_return` is okay, since it cannot be modified
+                    // again later. (`self` is consumed )
+                    should_return: self.should_return.clone(),
```

Comment:
I am concerted about it as `Promise` implements Drop:

https://github.com/near/near-sdk-rs/blob/cffda5c35204e7554151928cb5ee8ae9ab50f0cd/near-sdk/src/promise.rs#L576-L580

which calls `construct_recursively`:

https://github.com/near/near-sdk-rs/blob/cffda5c35204e7554151928cb5ee8ae9ab50f0cd/near-sdk/src/promise.rs#L564-L574

which calls `construct_recursively` on PromiseSingle, which calls the host functions to register a promise:

https://github.com/near/near-sdk-rs/blob/cffda5c35204e7554151928cb5ee8ae9ab50f0cd/near-sdk/src/promise.rs#L150-L154

Thus, cloning self will result in 1+N calls to the `self` promise it seems like.